README.txt
SPP - Synthetic Packet Pairs - 0.4 - Readme

  1. Overview

        Regular and frequent measurement of round trip time (RTT) between points
        on the Internet is becoming increasingly important for a range of highly
        interactive real-time applications. Active probing techniques are
        possible but problematic. The extra packet traffic imposed by active
        probes along a network path can modify the behaviour of the network
        under test. In addition, estimated RTT results may be misleading if the
        network handles active probe packets differently to regular IP packets.

        In contrast, Synthetic Packet Pairs (SPP [1][2]) provides frequently 
        updated RTT estimates using IP traffic already present in the network. 
        SPP estimates the RTT between two measurement points without requiring
        precise time synchronisation between each point. SPP estimates the RTT 
        experienced by an application's traffic without needing modifications to
        the application itself or the routers along the path. In addition, SPP 
        works with applications that do not exhibit symmetric client-server 
        packet exchanges (such as many online multiplayer games) and 
        applications generating IP multicast traffic.

        This software is a flexible, standalone packet processor that implements
        the SPP algorithm. RTT calculations can be generated from saved PCAP
        format files or local or remote interfaces in real time.

  2. Installation

        Please see INSTALL.txt found in this directory.


  3. Definitions

        RTT is the round trip time that Target Traffic experiences between two 
        Measurement Points.

        A Measurement Point is an interface (or tap on the network) at which
        Target Traffic is observed. The two measurement points are REF (reference)
        and MON (monitor).
        
        Target Traffic consists of a user nominated subset of traffic made up of
        packets passing REF on the way towards MON, and packets flowing past MON
        on the way towards REF.
        
        SPP identifies target traffic using two IP addresses. The REF point IP address
        is the source address of target traffic packets heading in the REF to MON
        direction, while the MON point IP address represents a source of target
        traffic packets heading in the MON to REF direction.
        
        Packet pairs are constructed by pairing up a packet seen heading from REF
        to MON with the next packet seen heading the other way from MON to REF.


  4. Operation

        SPP estimates RTT between two measurement points along a network path.
        Traffic is observed at both measurement points, and the RTT
        between the two measurement points is estimated from pairs of
        packets seen travelling in each direction.

        Data can be collected from measurement points in two ways. The simplest
        method is for the user to collect data at both monitor points manually
        before running SPP. This can be achieved using tcpdump on the
        measurement point interface. SPP can then read these PCAP files to
        estimate RTT. Of course this method can only provide retrospective
        RTT estimates.

        Alternatively, when near real time RTT estimates are required, SPP can
        passively monitor live network interfaces. Generally, the two
        measurement points are geographically separate, requiring SPP to be run
        on two separate hosts. The most common setup for near real time
        measurements involves an 'SPP master' and 'SPP slave'. At one measurement
        point, SPP is run as the master. At this point, SPP actually processes
        the data and outputs RTT estimates. At the other measurement point, SPP
        is run as a slave. It simply observes data and sends relevant 
        information to the SPP master. The relevant information in conveyed with
        SPP Sample Frames (SSF). These are UDP packets which contain enough 
        information to reconstruct IDs of observed packets and their 
        corresponding timestamps.

        (NOTE: Live remote monitoring is still experimental and may have bugs)

        In this implementation, target traffic is specified as packets flowing
        between two hosts. IP addresses of these hosts must be entered on
        the command line. The path between these two hosts must pass by the two
        measurement points. It may be that the measurement points reside on the
        hosts generating the target traffic, but this is not required.

  5. SPP Command Syntax and Options

        USAGE: File processing:
        spp -a address -A address -f file -F file [-# <hashcode> |-p|-c|-m|-b|-O|-P]


        USAGE: Live processing master:
        spp -a address -A address (-i interface | -r host )
        ( -I interface | -R slave_host ) [-# <hashcode> |-g usec |-p|-c|-m|-b|-O|-P]

        USAGE: Remote slave:
        spp  -a  address -A address -s master_host -I interface 
        [ -# <hashcode> | -g usec | -l no.bytes | -t seconds ]


      General Options:
        -a IP address at the reference point
        -A IP address at the monitor point
        -n Natted IP address of the reference point
        -N Natted IP address of the monitor point
        -s Put into slave mode and send SSF to specified host
        -v Verbosity Level - see man page
        -d T Delta Maximum (seconds) before autotuning starts or if autotuning disabled with '-T' (default: 60)
        -o Offset in seconds of the monitor point with respect to the reference point
        -G Search interval in number of packets (default: 10000(file)/500(live))
        -P Enable pcap/bpf filtering (only accept DLT_EN10MB-framed packets where IP addresses match)
        -H Specify hash function used for packet ID generation, must be "crc32" or "crc64" (default: crc64)
        -T Disable autotuning of T Delta using actual clock offset estimates

       Source options:
        -f File to be read for the reference point (PCAP format)
        -F File to be read for the monitor point (PCAP format)
        -i Reference point live capture interface
        -I Monitor point live capture interface

       Network Options:
        -l Length of remotely measured timestamps in bytes (default: 2)
        -g Granularity of remotely measured timestamps in microseconds (default: 100)
        -t Timeout - max time in seconds between updates from slave

       Output options:
        -p Output 'Server Processing Times'
        -c Output 'Pair Count'
        -m Calculate timestamps from monitor point clock
        -b Use the timestamp of the first packet in the pair for the pair timestamp
        -O Append 'fake' (uncorrected) OWD in each direction to each RTT line, [OWDref2mon OWDmon2ref]

        Packet Matching Options:
        -# <hashcode> (default: 63)
        The # option maybe used to set which fields are used in the packet matching process.
        The value of <hashcode> is the total of all the required field IDs as listed below:
        
        IP fields:
                        1 Source Address
                        2 Destination Address
                        4 Protocol
                        8 Identification
        TCP/UDP fields:
                        16 Source Port
                        32 Destination Port
        TCP fields:
                        64 Sequence Number
                        128 Acknowledgement Number
                        256 Data offset, flags, window size
                        512 Checksum, urgent pointer
                        8192 Up to 12 bytes of TCP payload (limited by packet length)
                        16384 All TCP Options bytes (if present)
        UDP Fields:
                        1024 Length, checksum
                        2048 Up to 12 bytes UDP data (limited by packet length)
        Not UDP/TCP:
                        4096 Up to 20 bytes after IP header (limited by packet length)


        NOTE: IP addresses will not be used for hashing when NAT is in use.

  6. Packet identification and clocks 
  
   6.1 Packet identification
   
        A crucial step in pairing packets is identifying each packet seen at REF
        with the same packet seen at MON (separately in each direction). SPP does
        this by generating a per-packet hash across a number of fields in the
        IP header, transport protocol header and/or payload. The '-#' option
        controls what specific combination of fields are used to generate the hash.
        
        Reliable disambiguation of packets requires hashing over fields that
        vary from one packet to the next, yet are invariant between REF and MON
        (not altered by network devices along the path). Some problematic scenarios
        include NAT (where IP addresses are not invariant along a path, and
        TCP/UDP ports may also be altered) and TCP sequence number remapping
        (observed being performed by certain 'security' middleboxes).
        
        When SPP was first developed, the IP.ID field was often unique for
        every IP packet emitted by a sender, and could be relied on to disambiguate
        retransmissions of higher later segments. However, RFC 6864 has formalised
        the notion that IP.ID need only be unique for fragments of a larger IP packet.
        
        The TCP Option bytes are useful for disambiguating TCP packets (including
        retransmissions) where the underlying connections have negotiated (and
        correctly use) the Time Stamp option. In such cases, retransmissions will
        always differ by their TSval field.
        
        If you find spp is generating implausibly high RTTs from time to time (such as
        when the hash fails to disambiguate a retransmitted TCP segment at MON from its
        orignal seen at REF), use a custom "-# <hashcode>" to hash over additional fields.

        If you find spp is not generating RTT estimates, use a custom "-# <hashcode>" to
        hash over fewer fields. (For example, don't hash over TCP sequence or acknowledgement
        numbers if a middle-box is rewriting these fields mid-path. Otherwise spp will
        fail to match a packet seen at REF with the same packet seen at MON.)

        Until version 0.3.6 spp used CRC32 as hash function. Since version 0.4
        CRC64 is used by default to reduce the collision probability. The -H option
        can be used to select CRC32 instead.
        
   6.2 Clock Synchronisation

        The SPP algorithm does not strictly require clocks at REF and MON to be
        synchronised. Nevertheless, this SPP implementation applies a practical

Changelog.txt
Version 0.4.2
- Fixed globals in master and slave (had a few non-static globals with same name)

Version 0.4.1
- Fixed bug that included random bytes in the hash input used to generate the packet id
  in cases where the captured bytes did not include the TCP option bytes or the first
  12 bytes for TCP/UDP payload. Now for the TCP options, the UDP/TCP payloads and data
  for non UDP/TCP, the capture length (snap size) will limit the hash input.

Version 0.4
- Added pointer to http://caia.swin.edu.au/reports/130730A/CAIA-TR-130730A.pdf
- Fixed threading bug that causes segfault with newer pthread lib on Linux
- Directory bin is now created during make (if it doesn't exists)
- Added support for DLT_PPP frames in pcap files
- Create fake "ip_len" if missing (observed in some TSO'ed frames)
- Noted that IP.ID is no longer useful for disambiguating packets (RFC 6864)
- Added "-# 8192" to hash across first 12 bytes of TCP payload
- Added "-# 16384" to hash across all TCP Option bytes
- Added -O option to print 'fake' (uncorrected) OWD in each direction
- Various documentation tweaks (README and man page source)
- Change default hash function for packet id generation to crc64 and added
- Added -H option to select hash function (crc32 and crc64)
- Added autotuning of delta_t_max based on actual observed clock offset
  (disabled with -T option)

Version 0.3.6
- Fixed bug that caused SPP to fail on FreeBSD 10.1
- Fixed a number of potential issues / compiler warnings

Version 0.3.5
- Fixed copyright notice

Version 0.3.4
- Fixed bug that stopped processing before the end of traces was reached
- Fixed bug that caused the output of negative timestamps (instead of zero timestamps)
- Fixed bug that caused hang in live mode after pressing ctrl-c on Linux
- Updated makefile and man page

Version 0.3.3
- Fixed deadlock scenario(s)
- Fixed incorrect timestamps for times with odd seconds value
- Fixed pthread mutex bug
- Added option to change number of packets searched before giving up (and increased default value)
- Added missing TAILQ macros for Linux compatibility
- Added more optional fields for packet ID computation

Version 0.3.2
- Fixed live capture
- Updated man page and documentation

Version 0.3.1
- Fixed bugs that would cause crashes on 64bit systems

Version 0.3.0
- Added support for NAT (Thanks to David Hayes)
- Bug fix 
  (Previously when working with large files, it was possible that a large
  number of measurements would be missing due to incorrect memory management.)  

Version 0.2.0
- Multiple hash lengths
- Multiple granularity settings for remote timestamps
- Maximum time between remote updates limit

Version 0.1.7
- Live interface capture
- Variable hashing
- Proof of concept remote capture

Version 0.1.6
- First usable multithreading version
- Calculates hashes on id, protocol, src dst addresses.


src/instance.c
/*

    Copyright 2008--2013, Centre for Advanced Internet Architectures,
    Swinburne University of Technology, http://caia.swin.edu.au
 
    Author: Amiel Heyde, amiel@swin.edu.au
 
    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License version 2 as 
    published by the Free Software Foundation.
 
    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA

    $Id: instance.c 176 2015-05-21 00:35:27Z szander $
 
    Note to self - Changes to 0.3.1:
    Used explicit casting to enable compilation on 64bit: *(struct in_addr *)& 
    See http://www.gidforums.com/t-7865.html for details (in_addr_t vs in_addr)

*/

#include <pcap.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <sys/queue.h>


#include <sys/types.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <netinet/if_ether.h>
#include <netinet/tcp.h>
#include <pthread.h>
#include <unistd.h>

#include "spptool.h"
#include "crc32.h"
#include "crc64.h"
#include "timeval.h"
#include "instance.h"
#include "master.h"
#include "slave.h"

//######## VARIABLES #########//
PUBLIC monitor_point_t mp[MP_COUNT];

PRIVATE char errbuf[PCAP_ERRBUF_SIZE];
PUBLIC in_addr_t addr[2];
PUBLIC in_addr_t nat_addr[2];

extern int hash_fields;
extern HASH_FUNCTION hash_function;
extern int finished;
extern int options;
extern int delta_t_max;
extern int verbosity;
extern unsigned int sec_offset;

//######## PROTOTYPES ########//

PUBLIC void mpoint_load(monitor_point_t * mpoint, const mp_type_t type, const char * name, mp_id_t id);
PUBLIC void instance_unload(mp_id_t id);

PRIVATE instance_t* assembleInstance(const struct pcap_pkthdr *pcap_hdr, const struct ip *ip_hdr, unsigned short ip_caplen,
                                     direction_t direction);
PUBLIC void createInstance(u_char *args, const struct pcap_pkthdr *pcap_hdr, const u_char *pkt);
PUBLIC void * createInstances(void * args);
PRIVATE void setPcapFilter(monitor_point_t * mpoint, char * filter_string);
PUBLIC uint64_t getHash(const struct ip *ip_hdr, unsigned short ip_caplen);

//######## FUNCTIONS #########//

/*
 * Initialises the monitor point struct. Called once for MON and once for REF.
 * */
PUBLIC void mpoint_load(monitor_point_t * mpoint, const mp_type_t type, const char * name, mp_id_t id) 
{
  mpoint->id = id;

  mpoint->q_size[IN] = 0;
  mpoint->q_size[OUT] = 0;
  mpoint->type = type;
  switch(type) {
    
    case live:
            
                if((mpoint->dev = pcap_open_live(name, BUFSIZ, 1, 500, errbuf)) == NULL) {
                 printf("PCAP error: %s\n", errbuf); 
                 exit(-1);
                }
                //Possible auto IP discovery from interface
               // if(pcap_lookupnet(pcap_lookupdev(errbuf), netp, errbuf)) { 
               //  printf("PCAP error: %s\n", errbuf); 
               //  exit(-1);
               // }
                break;
    case file:
                // Open PCAP file for reading
                if((mpoint->dev = pcap_open_offline(name, errbuf)) == NULL) {
                 printf("Error opening file: %s\n", errbuf); 
                  exit(-2);
                }
                mpoint->byte_order_swapped = pcap_is_swapped(mpoint->dev);
                break;
    case remote:
                loadMaster(mpoint, name);
                break;
    default:
                break;
  }


  TAILQ_INIT(&mpoint->instance_q[IN]);
  TAILQ_INIT(&mpoint->instance_q[OUT]);

  pthread_mutex_init(&mpoint->thresh_mutex, NULL);
  pthread_cond_init(&mpoint->thresh_cond, NULL);

  pthread_mutex_init(&mpoint->q_mutex[IN], NULL);
  pthread_mutex_init(&mpoint->q_mutex[OUT], NULL);
 
  pthread_mutex_init(&mpoint->q_size_mutex[OUT], NULL);
  pthread_mutex_init(&mpoint->q_size_mutex[IN], NULL);

  mpoint->finished = 0;
  if(mpoint->type != remote) {
      mpoint->datalink_type = pcap_datalink(mpoint->dev);
        if(verbosity & 32) printf("Monitor point id %d has datalink type: %d\n", mpoint->id, mpoint->datalink_type);
  }
}


/* 
 * Constructs the PCAP filter string. 
 * Spawns the input reader threads 
 * */
PUBLIC void mpoint_start(monitor_point_t * mpoint) {

 if(options & use_pcap_filter) {
   if(mpoint->type != remote) {
      char filter_string[90];
      char buf[8];
      strcpy(filter_string, "host ");
      strcat(filter_string, inet_ntoa(*(struct in_addr *)&addr[REF]));
      if (nat_addr[0] == 0 && nat_addr[1] == 0)
        strcat(filter_string, " and host ");
      else //one of the ends has NAT
        strcat(filter_string, " or host ");
      strcat(filter_string, (char*)inet_ntoa(*(struct in_addr *)&addr[MON]));
      strcat(filter_string, " and !icmp[icmptype]==3 and !port ");
      sprintf(buf, "%d", PORT);
      strcat(filter_string, buf);
          if(verbosity & 32) printf("PCAP filter string: %s\n", filter_string);
      setPcapFilter(mpoint, filter_string);
    }
 }
  switch(mpoint->type) {
    case file:
    case live:
                pthread_create(&mpoint->thread, NULL, &createInstances, (void*)mpoint);
                if(verbosity & 2) printf("INFO: Thread started\n");
                break;
    case remote:
                pthread_create(&mpoint->thread, NULL, &runMaster, (void*)mpoint);
                break;
    default:
                break;
  }



}

/*
 * Shuts down pcap, frees up the TAILQ lists
 * */
PUBLIC void mpoint_unload(monitor_point_t * mpoint) {

  switch(mpoint->type) {
    case live:
    case file:
                // Close pcap device
                pcap_close(mpoint->dev);
                break;
    case remote:
                break;
    default:
                break;
  }

  //CLEAN BOTH INSTANCE QUEUES
  direction_t direction;
  instance_t *ins, *ins_next;


  for(direction = IN; direction <= OUT; direction++) {
    pthread_mutex_lock(&mpoint->q_mutex[direction]);
    ins = TAILQ_FIRST(&mpoint->instance_q[direction]);
    while (ins != NULL) {
      ins_next = TAILQ_NEXT(ins, entries);
      free(ins);
      ins = ins_next;
    }
    pthread_mutex_unlock(&mpoint->q_mutex[direction]);
  }
}

PRIVATE instance_t* assembleInstance(const struct pcap_pkthdr *pcap_hdr, const struct ip *ip_hdr, unsigned short ip_caplen,
                                     direction_t direction){

  instance_t * ins = malloc(sizeof(instance_t));
  //printf("))) Malloc to ptr: %u\n", ins);
  ins->pkt_id = getHash(ip_hdr, ip_caplen);

  ins->ts = pcap_hdr->ts;                                                                     // Store instance timestamp
  return ins;
}

/*
 * Creates hash of a packet that is used for the pkt_id field. 
 *
 * The pkt_id is crucial for determining whether a packet seen at REF
 * has also been seen at MON. We rely on hashing together one or more fields
 * that are expected to be vary from packet to packet but also be invariant along
 * the path between REF and MON. The global 'hash_fields' holds a bitmask
 * indicating which fields should be included in the hash calculation.
 * 
 * In SPP <= 0.3.6 we assumed IP.ID (Identification) field would be non-zero and
 * unique per packet emitted by a given source (at least for time periods longer than
 * a handful of RTTs). However, RFC 6864 (Feb 2013) officially deprecated this use of
 * IP.ID field, and it is now a largely unreliable mechanism for disambiguating packets
 * that might otherwise look the same based on other header fields.
 * 
 * Some other fields may or may not be invariant along a path. Certain middleboxes have
 * been observed in the wild actually twiddling with the raw TCP sequence numbers,
 * ensuring they wont match between REF and MON.
 * 
 * */
PUBLIC uint64_t getHash(const struct ip *ip_hdr, unsigned short ip_caplen) {

  unsigned int hash_offset = 0;
  unsigned char hash_data[HASH_DATA_LENGTH + 1];
  u_char * transport_hdr = ((u_char *)ip_hdr + (ip_hdr->ip_hl * 4));
  struct tcphdr *tcp_hdr = (struct tcphdr *)transport_hdr;
 
  //printf("assembling\n");
  if (nat_addr[0] == 0 && nat_addr[1] == 0) {	// Do not include IP addresses in hash when running through NAT (This would cause no hashes to match)
    if(hash_fields & 1) {                 // Add Source address field
      memcpy((void*)(hash_data + hash_offset), (void*)&ip_hdr->ip_src, sizeof(ip_hdr->ip_src));
      hash_offset += sizeof(ip_hdr->ip_src);
    }
    if(hash_fields & 2) {                 // Add Destination address field 
      memcpy((void*)(hash_data + hash_offset), (void*)&ip_hdr->ip_dst, sizeof(ip_hdr->ip_dst));
      hash_offset += sizeof(ip_hdr->ip_dst);
    }
  }
  if(hash_fields & 4) {                   // Add Protocol
    memcpy((void*)(hash_data + hash_offset), (void*)&ip_hdr->ip_p, sizeof(ip_hdr->ip_p));
    hash_offset += sizeof(ip_hdr->ip_p);
  }
  if(hash_fields & 8) {                   // Add Identification field
    memcpy((void*)(hash_data + hash_offset), (void*)&ip_hdr->ip_id, sizeof(ip_hdr->ip_id));
    hash_offset += sizeof(ip_hdr->ip_id);
  }


  if(ip_hdr->ip_p == 6 || ip_hdr->ip_p == 17) {  //We have a TCP or UDP packet
    if(hash_fields & 16) {                // Add TCP/UDP src
      memcpy((void*)(hash_data + hash_offset), (void*)transport_hdr, 2);
      hash_offset += 2;
    }
    if(hash_fields & 32) {                // Add TCP/UDP dst
      memcpy((void*)(hash_data + hash_offset), (void*)(transport_hdr + 2), 2);
     hash_offset += 2;
    }

    if(ip_hdr->ip_p == 6) {  //We have a TCP packet
      if(hash_fields & 64) {              // Add TCP Seq No
        memcpy((void*)(hash_data + hash_offset), (void*)(transport_hdr + 4), 4);
        hash_offset += 4;
      }
      if(hash_fields & 128) {             // Add TCP Ack No
        memcpy((void*)(hash_data + hash_offset), (void*)(transport_hdr + 8), 4);
        hash_offset += 4;
      }
      if(hash_fields & 256) {             // Add TCP data offset, flags, window size
	memcpy((void*)(hash_data + hash_offset), (void*)(transport_hdr + 12), 4);
	hash_offset += 4;
      }
      if(hash_fields & 512) {             // Add TCP Checksum, urgent pointer
        memcpy((void*)(hash_data + hash_offset), (void*)(transport_hdr + 16), 4);
        hash_offset += 4;
      }
      if(hash_fields & 8192) {            // Add up to first hash_bytes bytes of TCP payload
	unsigned short hash_bytes = 12;
	unsigned short tcp_data_start = tcp_hdr->th_off * 4;
	int tcp_data_len = ip_caplen - (ip_hdr->ip_hl * 4) - tcp_data_start;
        if (tcp_data_len > 0) {
          if (tcp_data_len < hash_bytes) {
	    hash_bytes = tcp_data_len;
	  }
	  memcpy((void*)(hash_data + hash_offset), (void*)(transport_hdr + tcp_data_start), hash_bytes);
	  hash_offset += hash_bytes;
        }
      }
      if(hash_fields & 16384) {           // Hash across all TCP options bytes if present
	unsigned short hash_bytes = tcp_hdr->th_off * 4 - 20; // Number of bytes of TCP options
        int tcp_optdata_len = ip_caplen - (ip_hdr->ip_hl * 4) - 20;
        if (tcp_optdata_len > 0) {
          if (tcp_optdata_len < hash_bytes) {
            hash_bytes = tcp_optdata_len;
          }
	  memcpy((void*)(hash_data + hash_offset), (void*)(transport_hdr + 20), hash_bytes);
	  hash_offset += hash_bytes;
	}
      }

    }
    else {  //Must be UDP
      if(hash_fields & 1024) {             // Add UDP length, checksum 
        memcpy((void*)(hash_data + hash_offset), (void*)(transport_hdr + 4), 4);
        hash_offset += 4;
      }
      if(hash_fields & 2048) {             // Add UDP payload (up to 12 bytes)
        unsigned short hash_bytes = 12;
	int data_len = ip_caplen - 8;
        if (data_len > 0) {
	  if (data_len < hash_bytes) {
	    hash_bytes = data_len;
	  }
	  memcpy((void*)(hash_data + hash_offset), (void*)(transport_hdr + 8), hash_bytes);	      	
	  hash_offset += hash_bytes;
        }
      }

    }
  } else {
	if(hash_fields & 4096) {
		// If not TCP or UDP add up to first 20 bytes past IP header
                unsigned short hash_bytes = 20;
		int ip_data_len = ip_caplen - ip_hdr->ip_hl * 4;
                if (ip_data_len > 0) {
		  if (ip_data_len < hash_bytes) {
		    hash_bytes = ip_data_len;
		  }
		  memcpy((void*)(hash_data + hash_offset), (void*)transport_hdr, hash_bytes);
		  hash_offset += hash_bytes;
                }
	}
  }

  return hash_function(0, hash_data, hash_offset);
}

/*
 * This is spawned inside its own thread. It is the "read the input stream" 
 * thread. One is created for each input stream (ie, one for each of REF 
 * and MON)
 * */
PUBLIC void * createInstances(void * args) {
    monitor_point_t * mpoint = (monitor_point_t *)args;

    pcap_loop(mpoint->dev, -1, createInstance, (u_char*)mpoint);
    sleep(1);         //Leave some time for all calculations to be finished before sending the telling the program to quit.
    mpoint->finished = 1;

    pthread_exit(NULL);
}

/*
 * Reads a single packet in and adds it to the incoming queue to be processed
 * by a different thread. 
 *
 * */
PUBLIC void createInstance(u_char *args, const struct pcap_pkthdr *pcap_hdr, const u_char *pkt)
{

  direction_t direction;
  struct ip * ip_hdr;

  in_addr_t src_addr, dst_addr;
  instance_t * ins;

  struct ether_header * eth_hdr = (struct ether_header *)pkt;
  monitor_point_t * mpoint = (monitor_point_t *)args; 
  int pkt_err = 0;
  int link_hdr_len = 0;
  unsigned short ip_caplen;

  if(finished)pthread_exit(NULL);                     // Shut this thread down if we have been told to finish
  
  direction = -1;                                     // No direction by default (-1)
  
 switch(mpoint->datalink_type) {
    case DLT_EN10MB:                                  //printf("Found Ethernet\n");
                      if(eth_hdr->ether_type != 8 && eth_hdr->ether_type != 2048) { // If we are not carrying IP
                            if(verbosity & 32) printf("Skipping Ethernet frame not containing IPv4\n");
                        return;
                      }
                      link_hdr_len = sizeof(struct ether_header);
                      break;
    case DLT_LOOP:
    case DLT_NULL:                                    //printf("Found Null/Loop\n"); 

                      if(*(uint32_t*)pkt != 2) {
                            if(verbosity & 32) printf("Skipping Null/Loopback frame not containing IPv4\n");
                        return;
                      }
                      link_hdr_len = 4;
                      break;
    case DLT_LINUX_SLL:
		      //Changed by David Hayes to reflect the pcap man page
                      link_hdr_len = 16;
                      break;
    case DLT_PPP:
                      // Support PPP-encapsulated frames with/without HDLC encaps
		      if ((*pkt == 0xFF) && (*(pkt+1) == 0x03)) {
			link_hdr_len = 4;
		      } else {
			link_hdr_len = 2;
		      }
                      break;
    default:
                      printf("DataLink type not supported\n");
                      exit(EXIT_FAILURE);
    }

  ip_hdr = (struct ip *)(pkt + link_hdr_len);
  
  if(ip_hdr->ip_v != 4){
    if(verbosity & 32) printf("INFO: Skipping Packet: not IPv4\n");
    pkt_err = 1;
  }

  ip_caplen = pcap_hdr->caplen - link_hdr_len;
  if(ip_hdr->ip_len == 0){
    // Some instances of captured TSO'ed frames have been seen with ip_len=0,
    // so re-construct a fake a lower-bound IP packet length based on how many bytes
    // actually captured (may be used later during pkt_id generation)
    ip_hdr->ip_len = htons(ip_caplen);
  }

  if(mpoint->byte_order_swapped){
    src_addr = ntohl(ip_hdr->ip_src.s_addr);   //TEST THIS?
//    printf("%s", src_addr);
    dst_addr = ntohl(ip_hdr->ip_dst.s_addr);   //TEST THIS?
  } else {
    src_addr = ip_hdr->ip_src.s_addr;
    dst_addr = ip_hdr->ip_dst.s_addr;
  }


  if(verbosity & 32) {
    char addr_string[2][16];
    strncpy(addr_string[0], inet_ntoa(*(struct in_addr *)&src_addr), 16); // inet_ntoa() not always thread-safe?
    strncpy(addr_string[1], inet_ntoa(*(struct in_addr *)&dst_addr), 16); // inet_ntoa() not always thread-safe?
    printf("INFO: Next packet from monitor point %u: src %s, dst %s\n", 
           mpoint->id, addr_string[0], addr_string[1]);
  }

  if(src_addr == addr[0] && dst_addr == addr[1])                   // Determine if we have found an OUT packet
    direction = OUT;
  else if (src_addr == nat_addr[0] && dst_addr == addr[1])
    direction = OUT;
  else if (src_addr == addr[0] && dst_addr == nat_addr[1])
    direction = OUT;
  else if(src_addr == addr[1] && dst_addr == addr[0])              // Determine if we have found an IN packet
    direction = IN;
  else if(src_addr == addr[1] && dst_addr == nat_addr[0]) 
    direction = IN;
  else if(src_addr == nat_addr[1] && dst_addr == addr[0]) 
    direction = IN;
  else
    pkt_err = 1;
       

  if(pkt_err != 1) {
     
    if(options & run_slave) {
      sendHashes(pcap_hdr, ip_hdr, ip_caplen, direction);
      //printf("For debugging only \n");
    }
    else {
      ins = assembleInstance(pcap_hdr, ip_hdr, ip_caplen, direction);
      pthread_mutex_lock(&mpoint->q_mutex[direction]);
      TAILQ_INSERT_TAIL(&mpoint->instance_q[direction], ins, entries);       // Insert instance into appropriate queue
      pthread_mutex_unlock(&mpoint->q_mutex[direction]); 

      pthread_mutex_lock(&mpoint->q_size_mutex[direction]); //uncommented by David Hayes
      mpoint->q_size[direction]++;                         // Increment queue length
      //printf("DEBUGGING... \n");
      pthread_mutex_unlock(&mpoint->q_size_mutex[direction]);
      //printf("DEBUGGING 2... \n");
      
      if(verbosity & 16) {
        printf("INFO: Added %lu to mpoint %u instance_q[%u] - timestamp: %llu.%06llu\n", ins->pkt_id, mpoint->id, direction, 
               (unsigned long long)ins->ts.tv_sec, (unsigned long long)ins->ts.tv_usec);
      }
    }
  }
  else if(verbosity & 32) {
    printf("Packet discarded - (Not matching src and dst address)\n");
    char addr_string[2][16];
    strncpy(addr_string[0], inet_ntoa(*(struct in_addr *)&addr[0]), 16);
    strncpy(addr_string[1], inet_ntoa(*(struct in_addr *)&addr[1]), 16);
    printf("Src and dst should be %s or %s\n", addr_string[0], addr_string[1]);
    
  }

  // Check to see if the processing queues are full. If they are, wait a little bit
  // If we're doign a live or remote capture, do not wait - presumably, the frame
  // is not added to the queue if the queue is full.

  // sza: I removed the sleeping for reading from files. If we sleep here we slow 
  // down the analysis process. We rather use a bit more CPU to finish as quickly
  // as possible
#if 0
  if(mpoint->type == file) {
    pthread_mutex_lock(&mpoint->q_size_mutex[IN]);
    pthread_mutex_lock(&mpoint->q_size_mutex[OUT]);
    while(mpoint->q_size[IN] >= Q_MAX_LEN && mpoint->q_size[OUT] >= Q_MAX_LEN) {
      pthread_mutex_unlock(&mpoint->q_size_mutex[IN]);
      pthread_mutex_unlock(&mpoint->q_size_mutex[OUT]);
      if(verbosity & 2) printf("INFO: Monitor point %u sleeping\n", mpoint->id);
      usleep(10);
      if(finished)pthread_exit(NULL);
      pthread_mutex_lock(&mpoint->q_size_mutex[IN]);
      pthread_mutex_lock(&mpoint->q_size_mutex[OUT]);
    }
    pthread_mutex_unlock(&mpoint->q_size_mutex[IN]);
    pthread_mutex_unlock(&mpoint->q_size_mutex[OUT]);
  }
#endif
  if(finished)pthread_exit(NULL);

}

/*
 * Remove a certain packet pair from the yet-to-be-processed queue
 * */
PUBLIC void removeInstance(instance_t * instance, monitor_point_t * mpoint, direction_t direction) {
    //printf("))) About to free ptr: %u\n", instance);
    pthread_mutex_lock(&mpoint->q_mutex[direction]);
    TAILQ_REMOVE(&mpoint->instance_q[direction], instance, entries);
    //printf("%%% removeInstance: removed from Q\n");
    pthread_mutex_unlock(&mpoint->q_mutex[direction]);
      free(instance);

    //printf("%%% removeInstance: freed\n");
    pthread_mutex_lock(&mpoint->q_size_mutex[direction]);
    mpoint->q_size[direction]--;
    pthread_mutex_unlock(&mpoint->q_size_mutex[direction]);
    //printf("%%% removeInstance: done\n");
    //displayQueueSize();
}

/*
 * Prune packets older than delta_t_max from the yet-to-be-processed queue.
 * Calls removeInstance() on each of the pruned packets
 * */
PUBLIC void removeOldInstances(monitor_point_t * mpoint, direction_t direction, struct timeval * cur_time) {
  int delta_t;
  instance_t *ins, *ins_tmp;
  pthread_mutex_lock(&mpoint->q_mutex[direction]);

  TAILQ_FOREACH_SAFE(ins, &mpoint->instance_q[direction], entries, ins_tmp) {
    pthread_mutex_unlock(&mpoint->q_mutex[direction]);
     delta_t = (cur_time->tv_sec - sec_offset) - ins->ts.tv_sec;

    if(delta_t  > (delta_t_max + 1)) {
      if(verbosity & 8) {
        printf("INFO: Removing old instance %lu\n", ins->pkt_id);
      }
      removeInstance(ins, mpoint, direction);
    }
    else {
      return;                    // No point searching any more as items are in chronological order
    }
    pthread_mutex_lock(&mpoint->q_mutex[direction]);
  }
  pthread_mutex_unlock(&mpoint->q_mutex[direction]);

}

/*
 * Applies the filter string, called from mpoint_start
 * */
PRIVATE void setPcapFilter(monitor_point_t * mpoint, char * filter_string) {
  struct bpf_program bpf_prog;
  
  /* Note: Using this filter implicitly limits us to DLT_EN10MB frames.
   * (https://www.tcpdump.org/manpages/pcap-filter.7.html)
   */

  if(pcap_compile(mpoint->dev, &bpf_prog, filter_string,0,1) == -1)
    printf("Error compiling BPF program: %s\n", pcap_geterr(mpoint->dev));

  if(pcap_setfilter(mpoint->dev, &bpf_prog) == -1)
    printf("Error setting BPF filter: %s\n", pcap_geterr(mpoint->dev));

  pcap_freecode(&bpf_prog);
}
